### PR TITLE
Added more test coverage for built-in token.

### PR DESCRIPTION
**Pull request by dmkozh**
### What

Now the tests should cover most of its functionality, with the notable exclusion of events.

Added tests cover:

- double initialization checks
- classic account auth with multisig
- invoker auth
- failed auth scenarios (wrong nonce/signer/args/fn)
- negative/large amounts in operations
- classic account balance boundaries
- fail scenarios for all the transfer ops

### Why

Testing built-in token

### Known limitations

Events tests are still missing

### CODE REVIEW

File: soroban-env-host/src/host.rs
Comment (sisuresh) [inline soroban-env-host/src/host.rs:1247]: What's the reason for this change and the similar one in `get_invoking_contract `? I'm hesitant to add test code to non-test functions. We would ideally set up the host the right way so these functions work as expected.

Comment (dmkozh) [reply soroban-env-host/src/host.rs:1247]: I'm open to suggestions, but we need to come up with a straightforward way to inject the values for tests as messing with the frames etc has nothing to do with the test logic.

Comment (leighmcculloch) [reply soroban-env-host/src/host.rs:1247]: >messing with the frames etc has nothing to do with the test logic.

We should lean into using the system as it is in tests. That's how we learn how the system really behaves. It's also how we learn what is wrong with the design of the current system. If the system is really difficult to use in tests, we might want to reconsider its design down the track, but we should lean into using it as fully and completely as possible.

Comment (dmkozh) [reply soroban-env-host/src/host.rs:1247]: I'm not sure what are the best practices for DI in Rust; what I'd do in C++ requires virtual functions which we don't have here.

Comment (dmkozh) [reply soroban-env-host/src/host.rs:1247]: > We should lean into using the system as it is in tests. 

Sure, in integration/system tests. I'd argue that unit tests should mostly cover the relevant logic (token interface here) and mock the details that aren't relevant to the tests (i.e. how exactly we get the invoker).

Comment (leighmcculloch) [reply soroban-env-host/src/host.rs:1247]: I defer to @graydon @sisuresh on what the most appropriate approach for the token contract living in the host. But I think we should lean into integration/system tests, because they're less likely to change as we refactor the internals.

Fwiw this is the approach we've taken with supporting contract developers writing tests in the SDK, because the chance of testing the actual behavior of the contract and not fabricated behavior is greater. I believe @jonjove would also encouraging because he'd seen how in stellar-core tests that fiddled with internals became brittle over time.

Comment (dmkozh) [reply soroban-env-host/src/host.rs:1247]: > because they're less likely to change as we refactor the internals.

The point here is to test the 'internals' in a way (the behavior of this specific interface), so the changes would be warranted if that behavior changes for whatever reason. Covering the invocation logic is not necessarily a goal, though I agree it would be nice to have a test that covers that too (without relying on the token's behavior much).

Anyway, I don't think we need to argue here - I agree that the current way is hacky and if we had a less hacky way to imitate the invoker, I'd use it. We could rather merge this temporarily and open an issue to clean this up, or figure out a proper way to do that and change the tests later. I would go with the latter only if it's quick enough, since I think it's better to have partial coverage than no coverage at all. I could also remove only the invoker tests temporarily.

Comment (sisuresh) [reply soroban-env-host/src/host.rs:1247]: So the issue here is that the frames are missing right? We have a similar issue open - https://github.com/stellar/rs-soroban-env/issues/485. I think this should be prioritized since we'll run into this issue more often as we test. 

I do have some questions though - 
1. Can you wrap the `TestToken` methods with `with_frame` calls?
2. I'm not sure if this should be used for the same reason Leigh and I mention above... but with testutils enabled, `call_n` pushes a [frame](https://github.com/stellar/rs-soroban-env/blob/91cf595508d475877f855619ad5baeb5f45a3885/soroban-env-host/src/host.rs#L794) onto the stack.

Comment (dmkozh) [reply soroban-env-host/src/host.rs:1247]: I saw that issue. However what we need to do for the tests is imitate calls by an arbitrary account and by an arbitrary contract (that doesn't have to actually exist), which I believe is not exactly what it is about.

> but with testutils enabled, call_n pushes a [frame](https://github.com/stellar/rs-soroban-env/blob/91cf595508d475877f855619ad5baeb5f45a3885/soroban-env-host/src/host.rs#L794) onto the stack.

It's a bit confusing, but `testutils` are not enabled for the host tests themselves. We could add `test` to the respective conditions, but at that point I'm not sure what are we winning here, given that concern is primarily about using test-only code. I'm also not sure if we cover anything additional (besides testutils... which is not what token tests should cover in the first place). I could try using test frames to reduce the amount of test-only code, but I'm not sure if that really solves the issue of DI/mocking.

Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Using the frames was much simpler than I thought, so I got rid of the test-invoker functions. This comes at a cost of marking `TestContract` frame as test, but it's probably still cleaner. PTAL

Comment (leighmcculloch) [reply soroban-env-host/src/host.rs]: > This comes at a cost of marking TestContract frame as test, but it's probably still cleaner.

Assuming you mean the following, I think that is fine. Generally we should always accompany `features = "testutils"` with `test` so that all test utilities are available in tests by default.
```rust
#[cfg(any(test, feature = "testutils"))]
TestContract(Hash, Symbol),
```
There are a couple reasons for this:
- If testutils are reasonable for consumers of our crate to use in their tests, they should be reasonable for us to use.
- It prevents us from needing to run tests twice once with testutils and once without, so it reduces build time and test time.

**Review by sisuresh — APPROVED** [review]

File: soroban-env-host/src/host.rs
Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Sounds good, I'll keep this in mind for the future changes.

File: soroban-env-host/src/test/token.rs
Comment (leighmcculloch) [inline soroban-env-host/src/test/token.rs:283]: @dmkozh This is a really neat idea! 👏🏻  I think we should add the same thing to the `Env` in `testutils` in the SDK, as there are some situations where in a test we want to be able to access something _as_ a contract, and we currently have no way of doing that.

File: soroban-env-host/src/test/token.rs
Comment (leighmcculloch) [inline soroban-env-host/src/test/token.rs:383]: 💡 Could we change the `.is_err()` to do an assertion on the `Err(value)`? Tests that check for the presence of an error can rot over time because an error may start occurring that has nothing to do with what the test is happening, but the test won't fail and alert us to that.

Comment (leighmcculloch) [reply soroban-env-host/src/test/token.rs:383]: Actually, I guess `HostError` is not comparable with itself because it doesn't implement `ParitalEq`, because deep down it contains `RawVal` values that are not comparable. 

@graydon This is another example of where `RawVal` lacking equality limits us from writing more specific tests.

File: soroban-env-host/src/test/token.rs
Comment (leighmcculloch) [inline soroban-env-host/src/test/token.rs:687]: 👏🏻 The comments are very helpful in these tests.

**Review by leighmcculloch — APPROVED** [review]
👏🏻 Really awesome to get comprehensive tests across this area. Thanks for adding narrative comments to the tests, it made the tests very understandable / accessible. Some comments inline, but nothing to do.

File: soroban-env-host/src/test/token.rs
Comment (dmkozh) [reply soroban-env-host/src/test/token.rs:383]: Sure, I've already updated the tests, the PR was just blocked by this PR and is blocked by https://github.com/stellar/rs-soroban-env/pull/532 (maybe I can include the tests there though).

I had to extract status manually, so I agree that some functionality for comparison would help. OTOH, without a meaningful debug representation it wouldn't be too useful. While writing these tests I realized that I would much rather convert host value to a regular value and use `assert_eq` on it (like I do here with e.g. `to_vec`), rather than doing an equality check on the host values with `==` and without a meaningful error message in case of a failure.

File: soroban-env-host/src/test/token.rs
Comment (dmkozh) [reply soroban-env-host/src/test/token.rs:687]: Thanks! I would we revisit the test framework though as these comments are a poor substitution for something like sections in catch2 in C++ (there is [rust-catch](https://github.com/guydunton/rust-catch), but not sure how stable it is).

**Pull request merged**
